feat: Add resizing for QTable Columns: Issue #5115

Beanminchild opened the pull request:
<!--
  Please make sure to read the Pull Request Guidelines:
  https://github.com/quasarframework/quasar/blob/dev/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- Update "[ ]" to "[x]" to check a box (space sensitive) -->

**What kind of change does this PR introduce?** <!-- Check at least one -->

- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** <!-- Check one -->

- [ ] Yes
- [X] No

<!-- If yes, please describe the impact and migration path for existing applications: -->

**The PR fulfills these requirements:**

- [X] It's submitted to the `dev` branch (or `v[X]` branch)
- [X] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on an Electron app
- [ ] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar/tree/dev/docs) <!-- for faster update click on "Suggest an edit on GitHub" at bottom of page --> or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [X] Addressing open issue  https://github.com/quasarframework/quasar/issues/5115 // This small update adds a prop for a user resizable columns for the Qtable component.

**Other information:**
I am excited to submit this pull request, which introduces a feature that has been highly requested by the community, including myself. This is my first contribution to the project, and I have designed the component to be highly customizable to accommodate various use cases.

I welcome any feedback or suggestions for improvements, and I am open to making any necessary adjustments to ensure it aligns with the project's standards and goals.

Thank you for considering my contribution!


// Example Photos from test app used to verify functionality

![image](https://github.com/user-attachments/assets/807484b0-6b3a-4f80-86ac-d4475b2dd2d0)
![image](https://github.com/user-attachments/assets/6ec111fd-0913-43e7-9a7e-fca14b4ae01f)

Example table with one resizable column:
![image](https://github.com/user-attachments/assets/bf528495-d5a7-4462-b469-90dca3af6426)

Beanminchild committed:
ZE-0 add resizing for QTable Columns

Beanminchild commented:
Can I get a review on this please?

pdanpdan commented:
You mixed
- 4 space indent with the existing 2 space indent
- options API with the existing composition API
- semicolons in SASS

This change would also require to add the description of the added things in the API description JSON

If you want a review on code please change the PR so that it matches the existing code.

Beanminchild commented:
I'll work on that thank you for the feedback :)

pdanpdan commented on ui/src/components/table/QTable.js:
```diff
@@ -43,6 +43,11 @@ export default createComponent({
       type: [ String, Function ],
       default: 'id'
     },
+    resizableCols: {
```
`resizableCols: Boolean,` is enough

maybe it would be nice to use the columns definition to specify if a column is resizable or not

Beanminchild replied:
Good Idea,  I'll work on adding that, as well as fixing the other things you mentioned over the next day or two.

Beanminchild replied:
Addressed needed fixes, still looking into adding functionality to specified columns ... didn't have enough time to get to it this morning

Beanminchild committed:
address required fixes

Beanminchild committed:
minor improvments

Beanminchild committed:
- use column definitions to determine if a column is resizable or not

- double clicking resizer brings column back to their default size

Beanminchild commented:
hey @pdanpdan  sorry to ask but could you tell me where  the api description json is? 

So far i cleaned up most of what you wanted and added ability to use column definitions to specify which columns are resizable.  Also made it so you can double click the grabbers to make the columns go back to their default size.

Still need to clean up code a bit and add the description of the new functionality to the api description json.

pdanpdan commented:
https://github.com/quasarframework/quasar/blob/dev/ui/src/components/table/QTable.json

Beanminchild commented:
> https://github.com/quasarframework/quasar/blob/dev/ui/src/components/table/QTable.json

thank you very much

Beanminchild committed:
-cleaned up code to match rest of project
-documented new feature in QTable.json

Beanminchild committed:
-improved specificity of QTable.json description for resizable-cols prop

TheMeanBeanMachine commented:
Two of the example tables i made to test the feature: 

[Resize_Example_QTable_Default.txt](https://github.com/user-attachments/files/19017796/Resize_Example_QTable_Default.txt)

[Resize_Example_QTable_Customized_Using_Col_Defs_1Row.txt](https://github.com/user-attachments/files/19017801/Resize_Example_QTable_Customized_Using_Col_Defs_1Row.txt)

pintaf commented:
Hello. Any news on this ? Looking forward to see it integrated

vadim-popa commented:
Hi. I'm looking forward too. @pdanpdan do you have any updates?

timbze commented:
I would like this feature too, but it's clearly not good enough yet to be reviewed. I did a little cleanup in my own branch here that you can review. https://github.com/timbze/quasar/commit/255381712e3a68bb87f1d8414d2efab97c828e38

One rule of PRs for good code is not to change any existing code that is not related to your PR. Your PR changes a lot of spacing, adding semi-colons etc. I know it's probably your IDE or AI that did it, but still you need to review it and not submit anything that is not related to your changes.

One of the reasons why I know this won't be accepted yet is the manual 150 width that you add on mount. I don't have enough time to figure out how to do it correctly but IMHO this is clearly not it. Likely you should not add `onMounted` at all.

rstoenescu commented:
Thanks for this, and sorry it sat for so long.

We're clearing out the PR backlog and consolidating overlapping work: we're tracking QTable column resizing in #18184, which is the fuller implementation and includes the docs changes.

Closing this one so the effort stays in one place. Thanks for putting the time in.
